From 1d1a1c241a0ea3a9ee80f1845c52c3703cb7e5d7 Mon Sep 17 00:00:00 2001
From: Sam Leffler <sleffler@chromium.org>
Date: Tue, 22 Nov 2011 15:48:59 -0800
Subject: [PATCH] CHROMIUMOS: wpa_supplicant: make bgscan_simple scan all
 channels

Correct the logic to form the frequency list for scan requests so that
it includes all supported channels.  We were wrongly copying the zero
sentinel to the front of the frequenty list causing the scan request
to be discarded.

BUG=none (possibly many but nothing specific)
TEST=manual:wpa_debug debug and ping the next hop gw so scans are aborted; walk around to trigger scanning and roaming and verify the scan list is always complete

Change-Id: I91c5b214e8a09635dcde49daa26fdbd14045f304
Reviewed-on: https://gerrit.chromium.org/gerrit/12052
Reviewed-by: Paul Stewart <pstew@chromium.org>
Commit-Ready: Sam Leffler <sleffler@chromium.org>
Tested-by: Sam Leffler <sleffler@chromium.org>
---
 wpa_supplicant/bgscan_simple.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/wpa_supplicant/bgscan_simple.c b/wpa_supplicant/bgscan_simple.c
index 6b04142..459c643 100644
--- a/wpa_supplicant/bgscan_simple.c
+++ b/wpa_supplicant/bgscan_simple.c
@@ -43,8 +43,7 @@ static int * bgscan_simple_get_freqs(struct bgscan_simple_data *data)
 	if (data->supp_freqs == NULL)
 		return NULL;
 
-	/* NB: n_supp_freqs has +1 for trailing zero */
-	freqs = os_malloc(data->n_supp_freqs * sizeof(int));
+	freqs = os_malloc((data->n_supp_freqs + 1) * sizeof(int));
 	if (freqs == NULL)
 		return NULL;
 
@@ -53,6 +52,7 @@ static int * bgscan_simple_get_freqs(struct bgscan_simple_data *data)
 		freqs[j++] = data->supp_freqs[i];
 	for (i = 0; i < data->freq_idx; i++)
 		freqs[j++] = data->supp_freqs[i];
+	freqs[j] = 0;		/* NB: terminator expected elsewhere */
 
 	return freqs;
 }
@@ -217,7 +217,7 @@ static void bgscan_simple_setup_freqs(struct wpa_supplicant *wpa_s,
 	if (freqs != NULL) {
 		/* TODO(sleffler) priority order freqs */
 		data->supp_freqs = freqs;
-		data->n_supp_freqs = count+1;	/* NB: +1 for terminating 0 */
+		data->n_supp_freqs = count;
 
 		log_freqs("Supported", freqs);
 	}
-- 
2.2.0.rc0.207.ga3a616c

